coalaip/models.py

```diff
@@ -0,0 +1,380 @@
+"""Low level models mirroring COALA IP's entity model
+
+Supports transformations of the entity into JSON, JSON-LD, and IPLD data
+formats, as well as creation and transferring through persistence layer
+plugin.
+"""
+
+from copy import copy
+from coalaip import context_urls
+from coalaip.exceptions import (
+    EntityError,
+    EntityDataError,
+    EntityNotYetPersistedError,
+    EntityPreviouslyCreatedError,
+)
+from coalaip.plugin import AbstractPlugin
+
+
+DEFAULT_LD_CONTEXT = [context_urls.COALAIP, context_urls.SCHEMA]
+
+
+class CoalaIpEntity:
+    """Base class of all COALA IP entity models.
+
+    Provides base functionality for all COALA IP entities, including
+    entity creation (:method:`~CoalaIpEntity.create`) and, if necessary,
+    retrival of their status (:method:`~CoalaIpEntity.get_status`), on
+    the backing persistence layer provided by the given ledger plugin.
+    """
+
+    def __init__(self, data, *, ctx=DEFAULT_LD_CONTEXT, entity_type, plugin):
+        """Initialize a :class:`~coalaip.models.CoalaIpEntity` instance.
+
+        INSTANTIATION NOTES
+        """
+
+        if not isinstance(plugin, AbstractPlugin):
+            raise TypeError(('A plugin subclassing '
+                             "'coalaip.plugin.AbstractPlugin' must be "
+                             'provided when instantiating a CoalaIp entity '
+                             'instance. Given a plugin of type '
+                             "'{}' instead.".format(type(plugin))))
+
+        if not isinstance(entity_type, str):
+            raise EntityDataError(('The entity type must be provided as a '
+                                   'string to the entity. '
+                                   'Got {} instead.').format(entity_type))
+
+        # FIXME: should I check that data is a dict?
```

Comment:
Highlighting this for review.

Thinking about immediately throwing if `data` isn't a dict.